File: infrastructure/ts-sdk/src/dapr.ts
/**
 * Dapr client with state management, caching (TTL + metadata), pub/sub,
 * service invocation, secrets, and cache-aside pattern.
 */

interface CacheEntry<T> {
  value: T;
  cachedAt: number;
  ttl: number;
}

export class DaprClient {
  private baseUrl: string;
  private stateStore: string;
  private cacheStore: string;
  private pubsubName: string;
  private localCache = new Map<string, CacheEntry<unknown>>();
  private readonly localCacheMaxSize = 1000;

  constructor(httpPort: number = 3500, stateStore: string = 'statestore', pubsubName: string = 'pubsub') {
    this.baseUrl = `http://localhost:${httpPort}/v1.0`;
    this.stateStore = stateStore;
    this.cacheStore = 'cachestore'; // Separate Dapr component for caching
    this.pubsubName = pubsubName;
  }

  async ping(): Promise<void> {
    const resp = await fetch(`${this.baseUrl}/healthz`);
    if (!resp.ok) throw new Error(`Dapr unhealthy: ${resp.status}`);
  }

  // ─── State Management (strong consistency) ───────────────────────────────────

  async saveState(key: string, value: unknown, etag?: string): Promise<void> {
    const item: Record<string, unknown> = { key, value };
    if (etag) item.etag = etag;
    item.options = { concurrency: 'first-write', consistency: 'strong' };
    await fetch(`${this.baseUrl}/state/${this.stateStore}`, {
      method: 'POST', headers: { 'Content-Type': 'application/json' }, body: JSON.stringify([item]),
    });
    // Invalidate local cache on write
    this.localCache.delete(key);
  }

  async getState<T = unknown>(key: string): Promise<{ value: T | null; etag: string }> {
    const resp = await fetch(`${this.baseUrl}/state/${this.stateStore}/${key}`);
    if (!resp.ok) return { value: null, etag: '' };
    const etag = resp.headers.get('ETag') || '';
    const value = await resp.json() as T;
    return { value, etag };
  }

  async deleteState(key: string): Promise<void> {
    await fetch(`${this.baseUrl}/state/${this.stateStore}/${key}`, { method: 'DELETE' });
    this.localCache.delete(key);
  }

  // ─── Caching Layer (TTL-based, eventual consistency) ─────────────────────────

  /** Cache a value with TTL using Dapr state store metadata. */
  async cacheSet(key: string, value: unknown, ttlSeconds: number = 300): Promise<void> {
    const item = {
      key: `cache:${key}`,
      value,
      metadata: { ttlInSeconds: String(ttlSeconds) },
    };
    await fetch(`${this.baseUrl}/state/${this.stateStore}`, {
      method: 'POST', headers: { 'Content-Type': 'application/json' }, body: JSON.stringify([item]),
    });
    // Update local cache
    this._setLocal(key, value, ttlSeconds);
  }

  /** Get a cached value (checks local first, then Dapr state store). */
  async cacheGet<T = unknown>(key: string): Promise<T | null> {
    // L1: Local in-memory cache
    const local = this._getLocal<T>(key);
    if (local !== null) return local;

    // L2: Dapr state store (Redis-backed)
    const resp = await fetch(`${this.baseUrl}/state/${this.stateStore}/cache:${key}`);
    if (!resp.ok) return null;
    const value = await resp.json() as T;
    // Populate local cache with remaining TTL estimate
    this._setLocal(key, value, 60);
    return value;
  }

  /** Delete a cached value. */
  async cacheDelete(key: string): Promise<void> {
    await fetch(`${this.baseUrl}/state/${this.stateStore}/cache:${key}`, { method: 'DELETE' });
    this.localCache.delete(key);
  }

  /** Cache-aside pattern: get from cache, or compute + cache on miss. */
  async cacheAside<T>(key: string, ttlSeconds: number, loader: () => Promise<T>): Promise<T> {
    const cached = await this.cacheGet<T>(key);
    if (cached !== null) return cached;
    const value = await loader();
    await this.cacheSet(key, value, ttlSeconds);
    return value;
  }

  private _setLocal(key: string, value: unknown, ttl: number): void {
    if (this.localCache.size >= this.localCacheMaxSize) {
      // Evict oldest entry
      const first = this.localCache.keys().next().value;
      if (first !== undefined) this.localCache.delete(first);
    }
    this.localCache.set(key, { value, cachedAt: Date.now(), ttl: ttl * 1000 });
  }

  private _getLocal<T>(key: string): T | null {
    const entry = this.localCache.get(key);
    if (!entry) return null;
    if (Date.now() - entry.cachedAt > entry.ttl) {
      this.localCache.delete(key);
      return null;
    }
    return entry.value as T;
  }

  // ─── Pub/Sub ─────────────────────────────────────────────────────────────────

  async publishEvent(topic: string, data: unknown): Promise<void> {
    await fetch(`${this.baseUrl}/publish/${this.pubsubName}/${topic}`, {
      method: 'POST', headers: { 'Content-Type': 'application/json' }, body: JSON.stringify(data),
    });
  }

  // ─── Service Invocation ──────────────────────────────────────────────────────

  async invokeService(appId: string, method: string, data?: unknown, httpMethod: string = 'POST'): Promise<unknown> {
    const opts: RequestInit = { method: httpMethod, headers: { 'Content-Type': 'application/json' } };
    if (data && httpMethod !== 'GET') opts.body = JSON.stringify(data);
    const resp = await fetch(`${this.baseUrl}/invoke/${appId}/method/${method}`, opts);
    if (!resp.ok) throw new Error(`Service invoke failed (${resp.status})`);
    const text = await resp.text();
    return text ? JSON.parse(text) : null;
  }

  // ─── Secrets ─────────────────────────────────────────────────────────────────

  async getSecret(secretStore: string, secretName: string): Promise<Record<string, string>> {
    const resp = await fetch(`${this.baseUrl}/secrets/${secretStore}/${secretName}`);
    if (!resp.ok) throw new Error(`Secret retrieval failed (${resp.status})`);
    return resp.json() as Promise<Record<string, string>>;
  }

  // ─── Domain Operations (KYC, Policy, Claims) ────────────────────────────────

  async saveKYCSession(sessionId: string, data: Record<string, unknown>): Promise<void> {
    await this.saveState(`kyc:session:${sessionId}`, { ...data, updated_at: new Date().toISOString() });
  }

  async getKYCSession(sessionId: string): Promise<Record<string, unknown> | null> {
    const { value } = await this.getState<Record<string, unknown>>(`kyc:session:${sessionId}`);
    return value;
  }

  async publishKYCEvent(eventType: string, customerId: string, data: Record<string, unknown>): Promise<void> {
    await this.publishEvent('kyc-events', { event_type: eventType, customer_id: customerId, data, timestamp: new Date().toISOString() });
  }

  async savePolicyState(policyId: string, state: Record<string, unknown>): Promise<void> {
    await this.saveState(`policy:${policyId}`, state);
  }

  async saveClaimState(claimId: string, state: Record<string, unknown>): Promise<void> {
    await this.saveState(`claim:${claimId}`, state);
  }

  /** Warm the Dapr cache with commonly-accessed state entries. */
  async warmCache(entries: Array<{ key: string; loader: () => Promise<unknown>; ttl: number }>): Promise<number> {
    let loaded = 0;
    for (const { key, loader, ttl } of entries) {
      try {
        const value = await loader();
        await this.cacheSet(key, value, ttl);
        loaded++;
      } catch (err) { console.error('[dapr] cache warmup entry failed:', err instanceof Error ? err.message : err); }
    }
    return loaded;
  }
}

File: infrastructure/rust-sdk/src/dapr.rs
//! Dapr client with state management, pub/sub, service invocation, and secrets.

use reqwest::Client;
use serde_json::json;
use std::time::Duration;

pub struct DaprClient {
    base_url: String,
    state_store: String,
    pubsub_name: String,
    client: Client,
}

impl DaprClient {
    pub fn new(http_port: u16) -> Self {
        Self {
            base_url: format!("http://localhost:{}/v1.0", http_port),
            state_store: "statestore".to_string(),
            pubsub_name: "pubsub".to_string(),
            client: Client::builder().timeout(Duration::from_secs(5)).build().unwrap_or_default(),
        }
    }

    pub async fn ping(&self) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
        self.client.get(format!("{}/healthz", self.base_url)).send().await?;
        Ok(())
    }

    pub async fn save_state(&self, key: &str, value: &serde_json::Value, etag: Option<&str>) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
        let mut item = json!({ "key": key, "value": value, "options": { "concurrency": "first-write", "consistency": "strong" } });
        if let Some(e) = etag { item["etag"] = json!(e); }
        self.client.post(format!("{}/state/{}", self.base_url, self.state_store))
            .json(&json!([item])).send().await?;
        Ok(())
    }

    pub async fn get_state(&self, key: &str) -> Result<(Option<serde_json::Value>, String), Box<dyn std::error::Error + Send + Sync>> {
        let resp = self.client.get(format!("{}/state/{}/{}", self.base_url, self.state_store, key)).send().await?;
        if !resp.status().is_success() { return Ok((None, String::new())); }
        let etag = resp.headers().get("ETag").and_then(|v| v.to_str().ok()).unwrap_or("").to_string();
        let value: serde_json::Value = resp.json().await?;
        Ok((Some(value), etag))
    }

    pub async fn delete_state(&self, key: &str) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
        self.client.delete(format!("{}/state/{}/{}", self.base_url, self.state_store, key)).send().await?;
        Ok(())
    }

    pub async fn publish_event(&self, topic: &str, data: &serde_json::Value) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
        self.client.post(format!("{}/publish/{}/{}", self.base_url, self.pubsub_name, topic))
            .json(data).send().await?;
        Ok(())
    }

    pub async fn invoke_service(&self, app_id: &str, method: &str, data: Option<&serde_json::Value>, http_method: &str) -> Result<serde_json::Value, Box<dyn std::error::Error + Send + Sync>> {
        let url = format!("{}/invoke/{}/method/{}", self.base_url, app_id, method);
        let resp = match http_method {
            "GET" => self.client.get(&url).send().await?,
            "PUT" => self.client.put(&url).json(&data.unwrap_or(&json!({}))).send().await?,
            "DELETE" => self.client.delete(&url).send().await?,
            _ => self.client.post(&url).json(&data.unwrap_or(&json!({}))).send().await?,
        };
        if !resp.status().is_success() {
            return Err(format!("Service invoke failed ({})", resp.status()).into());
        }
        let text = resp.text().await?;
        if text.is_empty() { return Ok(json!(null)); }
        Ok(serde_json::from_str(&text)?)
    }

    pub async fn get_secret(&self, secret_store: &str, secret_name: &str) -> Result<serde_json::Value, Box<dyn std::error::Error + Send + Sync>> {
        let resp = self.client.get(format!("{}/secrets/{}/{}", self.base_url, secret_store, secret_name)).send().await?;
        if !resp.status().is_success() {
            return Err(format!("Secret retrieval failed ({})", resp.status()).into());
        }
        Ok(resp.json().await?)
    }

    pub async fn save_kyc_session(&self, session_id: &str, data: &serde_json::Value) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
        let now = std::time::SystemTime::now().duration_since(std::time::UNIX_EPOCH).unwrap_or_default().as_secs();
        let mut val = data.clone();
        if let Some(obj) = val.as_object_mut() { obj.insert("updated_at".into(), json!(now)); }
        self.save_state(&format!("kyc:session:{}", session_id), &val, None).await
    }

    pub async fn get_kyc_session(&self, session_id: &str) -> Result<Option<serde_json::Value>, Box<dyn std::error::Error + Send + Sync>> {
        let (value, _) = self.get_state(&format!("kyc:session:{}", session_id)).await?;
        Ok(value)
    }

    pub async fn publish_kyc_event(&self, event_type: &str, customer_id: &str, data: &serde_json::Value) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
        let now = std::time::SystemTime::now().duration_since(std::time::UNIX_EPOCH).unwrap_or_default().as_secs();
        self.publish_event("kyc-events", &json!({ "event_type": event_type, "customer_id": customer_id, "data": data, "timestamp": now })).await
    }

    pub async fn save_policy_state(&self, policy_id: &str, state: &serde_json::Value) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
        self.save_state(&format!("policy:{}", policy_id), state, None).await
    }

    pub async fn save_claim_state(&self, claim_id: &str, state: &serde_json::Value) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
        self.save_state(&format!("claim:{}", claim_id), state, None).await
    }
}
